tracing-syslog/src/syslog.rs

```diff
@@ -0,0 +1,220 @@
+    LOG_LOCAL4 = libc::LOG_LOCAL4,
+    /// Reserved for local use.
+    LOG_LOCAL5 = libc::LOG_LOCAL5,
+    /// Reserved for local use.
+    LOG_LOCAL6 = libc::LOG_LOCAL6,
+    /// Reserved for local use.
+    LOG_LOCAL7 = libc::LOG_LOCAL7,
+}
+
+impl Default for Facility {
+    fn default() -> Self {
+        Self::LOG_USER
+    }
+}
+
+/// `syslog` severity.
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone)]
+#[repr(i32)]
+pub enum Severity {
+    /// System is unusable.
+    LOG_EMERG = libc::LOG_EMERG,
+    /// Action must be taken immediately.
+    LOG_ALERT = libc::LOG_ALERT,
+    /// Critical conditions.
+    LOG_CRIT = libc::LOG_CRIT,
+    /// Error conditions.
+    LOG_ERR = libc::LOG_ERR,
+    /// Warning conditions.
+    LOG_WARNING = libc::LOG_WARNING,
+    /// Normal, but significant, condition.
+    LOG_NOTICE = libc::LOG_NOTICE,
+    /// Informational message.
+    LOG_INFO = libc::LOG_INFO,
+    /// Debug-level message.
+    LOG_DEBUG = libc::LOG_DEBUG,
+}
+
+impl From<Level> for Severity {
+    fn from(level: Level) -> Self {
+        match level {
+            Level::ERROR => Self::LOG_ERR,
+            Level::WARN => Self::LOG_WARNING,
+            Level::INFO => Self::LOG_NOTICE,
+            Level::DEBUG => Self::LOG_INFO,
+            Level::TRACE => Self::LOG_DEBUG,
+        }
+    }
+}
+
+/// `syslog` priority.
+#[derive(Copy, Clone, Debug)]
+struct Priority(libc::c_int);
+
+impl Priority {
+    fn new(facility: Facility, level: Level) -> Self {
+        let severity = Severity::from(level);
+        Self((facility as libc::c_int) | (severity as libc::c_int))
+    }
+}
+
+fn syslog(priority: Priority, msg: &CStr) {
+    unsafe { libc::syslog(priority.0, "%s\0".as_ptr().cast(), msg.as_ptr()) }
+}
+
+/// Creates a `syslog`-backed [`tracing_core::Collect`].
+pub fn logger(
+    identity: impl Into<Cow<'static, CStr>>,
+    options: Options,
+    facility: Facility,
+) -> impl Collect {
+    Registry::default().with(Syslog::new(identity, options, facility))
+}
+
+/// [`tracing_subscriber::Subscribe`] that logs to `syslog`.
+pub struct Syslog {
+    /// Identity e.g. program name. Referenced by syslog, so we store it here to
+    /// ensure it lives until we are done logging.
+    #[allow(dead_code)]
+    identity: Cow<'static, CStr>,
+    facility: Facility,
+}
+
+impl Syslog {
+    /// Creates a new `syslog` logger.
+    pub fn new(
+        identity: impl Into<Cow<'static, CStr>>,
+        options: Options,
+        facility: Facility,
+    ) -> Self {
+        let identity = identity.into();
+        // SAFETY: identity will remain alive until the returned struct's fields
+        // are dropped, by which point `closelog` will have been called by the
+        // `Drop` implementation.
+        unsafe { libc::openlog(identity.as_ptr(), options.0, facility as libc::c_int) };
+        Syslog { identity, facility }
+    }
+}
+
+impl Drop for Syslog {
+    fn drop(&mut self) {
+        unsafe { libc::closelog() };
+    }
```

Comment:
Why's this safe? What invarients need to be upheld? If you can, can you add comments explaining why?

Reply:
I haven't decided yet how to best handle the case where multiple `Syslog`s are created (and thus `openlog()` is called multiple times). Recently created loggers will clobber the identity, facility, and options of previously created loggers, but what's worse is that when one is dropped and calls `closelog()` it will destroy the identity, facility, and options of the other ones. For these reasons, I think it might be best to have an `AtomicBool` or similar to keep track of whether or not a logger is currently initialized, and panic if a user tries to initialize multiple concurrently.